In [1]:
# The usual preamble
%matplotlib inline

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Make the graphs a bit prettier, and bigger
plt.style.use('default')

One of the main problems with messy data is: how do you know if it's messy or not?

We're going to use the NYC 311 service request dataset again here, since it's big and a bit unwieldy.

In [2]:
requests = pd.read_csv('../data/311-service-requests.csv')

/usr/local/Cellar/ipython/6.5.0/libexec/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


# 7.1 How do we know if it's messy? 

We're going to look at a few columns here. I know already that there are some problems with the zip code, so let's look at that first.
 
To get a sense for whether a column has problems, I usually use `.unique()` to look at all its values. If it's a numeric column, I'll instead plot a histogram to get a sense of the distribution.

When we look at the unique values in "Incident Zip", it quickly becomes clear that this is a mess.

Some of the problems:

* Some have been parsed as strings, and some as floats
* There are `nan`s 
* Some of the zip codes are `29616-0759` or `83`
* There are some N/A values that pandas didn't recognize, like 'N/A' and 'NO CLUE'

What we can do:

* Normalize 'N/A' and 'NO CLUE' into regular nan values
* Look at what's up with the 83, and decide what to do
* Make everything strings

In [3]:
requests['Incident Zip'].unique()

array([11432.0, 11378.0, 10032.0, 10023.0, 10027.0, 11372.0, 11419.0,
       11417.0, 10011.0, 11225.0, 11218.0, 10003.0, 10029.0, 10466.0,
       11219.0, 10025.0, 10310.0, 11236.0, nan, 10033.0, 11216.0, 10016.0,
       10305.0, 10312.0, 10026.0, 10309.0, 10036.0, 11433.0, 11235.0,
       11213.0, 11379.0, 11101.0, 10014.0, 11231.0, 11234.0, 10457.0,
       10459.0, 10465.0, 11207.0, 10002.0, 10034.0, 11233.0, 10453.0,
       10456.0, 10469.0, 11374.0, 11221.0, 11421.0, 11215.0, 10007.0,
       10019.0, 11205.0, 11418.0, 11369.0, 11249.0, 10005.0, 10009.0,
       11211.0, 11412.0, 10458.0, 11229.0, 10065.0, 10030.0, 11222.0,
       10024.0, 10013.0, 11420.0, 11365.0, 10012.0, 11214.0, 11212.0,
       10022.0, 11232.0, 11040.0, 11226.0, 10281.0, 11102.0, 11208.0,
       10001.0, 10472.0, 11414.0, 11223.0, 10040.0, 11220.0, 11373.0,
       11203.0, 11691.0, 11356.0, 10017.0, 10452.0, 10280.0, 11217.0,
       10031.0, 11201.0, 11358.0, 10128.0, 11423.0, 10039.0, 10010.0,
       11209.0,

# 7.2 Fixing the nan values and string/float confusion

We can pass a `na_values` option to `pd.read_csv` to clean this up a little bit. We can also specify that the type of Incident Zip is a string, not a float.

In [4]:
na_values = ['NO CLUE', 'N/A', '0']
requests = pd.read_csv('../data/311-service-requests.csv', na_values=na_values, dtype={'Incident Zip': str})

In [5]:
requests['Incident Zip'].unique()

array(['11432', '11378', '10032', '10023', '10027', '11372', '11419',
       '11417', '10011', '11225', '11218', '10003', '10029', '10466',
       '11219', '10025', '10310', '11236', nan, '10033', '11216', '10016',
       '10305', '10312', '10026', '10309', '10036', '11433', '11235',
       '11213', '11379', '11101', '10014', '11231', '11234', '10457',
       '10459', '10465', '11207', '10002', '10034', '11233', '10453',
       '10456', '10469', '11374', '11221', '11421', '11215', '10007',
       '10019', '11205', '11418', '11369', '11249', '10005', '10009',
       '11211', '11412', '10458', '11229', '10065', '10030', '11222',
       '10024', '10013', '11420', '11365', '10012', '11214', '11212',
       '10022', '11232', '11040', '11226', '10281', '11102', '11208',
       '10001', '10472', '11414', '11223', '10040', '11220', '11373',
       '11203', '11691', '11356', '10017', '10452', '10280', '11217',
       '10031', '11201', '11358', '10128', '11423', '10039', '10010',
       '11209',

# 7.3 What's up with the dashes?

In [6]:
rows_with_dashes = requests['Incident Zip'].str.contains('-').fillna(False)
len(requests[rows_with_dashes])

5

In [7]:
requests[rows_with_dashes]

Unique Key            Created Date             Closed Date Agency  \
29136    26550551  10/24/2013 06:16:34 PM                     NaN    DCA   
30939    26548831  10/24/2013 09:35:10 AM                     NaN    DCA   
70539    26488417  10/15/2013 03:40:33 PM                     NaN    TLC   
85821    26468296  10/10/2013 12:36:43 PM  10/26/2013 01:07:07 AM    DCA   
89304    26461137  10/09/2013 05:23:46 PM  10/25/2013 01:06:41 AM    DCA   

                          Agency Name      Complaint Type         Descriptor  \
29136  Department of Consumer Affairs  Consumer Complaint  False Advertising   
30939  Department of Consumer Affairs  Consumer Complaint         Harassment   
70539   Taxi and Limousine Commission      Taxi Complaint   Driver Complaint   
85821  Department of Consumer Affairs  Consumer Complaint      Debt Not Owed   
89304  Department of Consumer Affairs  Consumer Complaint         Harassment   

      Location Type Incident Zip        Incident Address   ...     \
29136           NaN   77092-2016   2700 EAST SELTICE WAY   ...      
30939           NaN   55164-0737          P.O. BOX 64437   ...      
70539        Street   11549-3650  365 HOFSTRA UNIVERSITY   ...      
85821           NaN   29616-0759            PO BOX 25759   ...      
89304           NaN   35209-3114         600 BEACON PKWY   ...      

      Bridge Highway Name Bridge Highway Direction Road Ramp  \
29136                 NaN                      NaN       NaN   
30939                 NaN                      NaN       NaN   
70539                 NaN                      NaN       NaN   
85821                 NaN                      NaN       NaN   
89304                 NaN                      NaN       NaN   

      Bridge Highway Segment Garage Lot Name Ferry Direction  \
29136                    NaN             NaN             NaN   
30939                    NaN             NaN             NaN   
70539                    NaN             NaN             NaN   
85821                    NaN             NaN             NaN   
89304                    NaN             NaN             NaN   

      Ferry Terminal Name Latitude Longitude Location  
29136                 NaN      NaN       NaN      NaN  
30939                 NaN      NaN       NaN      NaN  
70539                 NaN      NaN       NaN      NaN  
85821                 NaN      NaN       NaN      NaN  
89304                 NaN      NaN       NaN      NaN  

[5 rows x 52 columns]

I thought these were missing data and originally deleted them like this:

`requests['Incident Zip'][rows_with_dashes] = np.nan`

But then my friend Dave pointed out that 9-digit zip codes are normal. Let's look at all the zip codes with more than 5 digits, make sure they're okay, and then truncate them.

In [8]:
long_zip_codes = requests['Incident Zip'].str.len() > 5
requests['Incident Zip'][long_zip_codes].unique()

array(['77092-2016', '55164-0737', '000000', '11549-3650', '29616-0759',
       '35209-3114'], dtype=object)

Those all look okay to truncate to me.

In [9]:
requests['Incident Zip'] = requests['Incident Zip'].str.slice(0, 5)

Done.

Earlier I thought 00083 was a broken zip code, but turns out Central Park's zip code 00083! Shows what I know. I'm still concerned about the 00000 zip codes, though: let's look at that. 

In [10]:
requests[requests['Incident Zip'] == '00000']

Unique Key            Created Date Closed Date Agency  \
42600    26529313  10/22/2013 02:51:06 PM         NaN    TLC   
60843    26507389  10/17/2013 05:48:44 PM         NaN    TLC   

                         Agency Name  Complaint Type        Descriptor  \
42600  Taxi and Limousine Commission  Taxi Complaint  Driver Complaint   
60843  Taxi and Limousine Commission  Taxi Complaint  Driver Complaint   

      Location Type Incident Zip  Incident Address   ...     \
42600           NaN        00000           EWR EWR   ...      
60843        Street        00000  1 NEWARK AIRPORT   ...      

      Bridge Highway Name Bridge Highway Direction Road Ramp  \
42600                 NaN                      NaN       NaN   
60843                 NaN                      NaN       NaN   

      Bridge Highway Segment Garage Lot Name Ferry Direction  \
42600                    NaN             NaN             NaN   
60843                    NaN             NaN             NaN   

      Ferry Terminal Name Latitude Longitude Location  
42600                 NaN      NaN       NaN      NaN  
60843                 NaN      NaN       NaN      NaN  

[2 rows x 52 columns]

This looks bad to me. Let's set these to nan.

In [11]:
zero_zips = requests['Incident Zip'] == '00000'
requests.loc[zero_zips, 'Incident Zip'] = np.nan

Great. Let's see where we are now:

In [12]:
unique_zips = requests['Incident Zip'].sort_values().unique()
unique_zips

array(['00083', '02061', '06901', '07020', '07087', '07093', '07109',
       '07114', '07201', '07208', '07306', '07604', '08807', '10000',
       '10001', '10002', '10003', '10004', '10005', '10006', '10007',
       '10009', '10010', '10011', '10012', '10013', '10014', '10016',
       '10017', '10018', '10019', '10020', '10021', '10022', '10023',
       '10024', '10025', '10026', '10027', '10028', '10029', '10030',
       '10031', '10032', '10033', '10034', '10035', '10036', '10037',
       '10038', '10039', '10040', '10044', '10048', '10065', '10069',
       '10075', '10103', '10107', '10112', '10119', '10128', '10129',
       '10153', '10162', '10280', '10281', '10282', '10301', '10302',
       '10303', '10304', '10305', '10306', '10307', '10308', '10309',
       '10310', '10312', '10314', '10451', '10452', '10453', '10454',
       '10455', '10456', '10457', '10458', '10459', '10460', '10461',
       '10462', '10463', '10464', '10465', '10466', '10467', '10468',
       '10469', '104

Amazing! This is much cleaner. There's something a bit weird here, though -- I looked up 77056 on Google maps, and that's in Texas.

Let's take a closer look:

In [13]:
zips = requests['Incident Zip']
# Let's say the zips starting with '0' and '1' are okay, for now. (this isn't actually true -- 13221 is in Syracuse, and why?)
is_close = zips.str.startswith('0') | zips.str.startswith('1')
# There are a bunch of NaNs, but we're not interested in them right now, so we'll say they're False
is_far = ~(is_close) & zips.notnull()

In [14]:
zips[is_far]

12102    77056
13450    70711
29136    77092
30939    55164
44008    90010
47048    23541
57636    92123
71001    92123
71834    23502
80573    61702
85821    29616
89304    35209
94201    41042
Name: Incident Zip, dtype: object

In [15]:
requests[is_far][['Incident Zip', 'Descriptor', 'City']].sort_values('Incident Zip')

Incident Zip         Descriptor         City
71834        23502         Harassment      NORFOLK
47048        23541         Harassment      NORFOLK
85821        29616      Debt Not Owed   GREENVILLE
89304        35209         Harassment   BIRMINGHAM
94201        41042         Harassment     FLORENCE
30939        55164         Harassment     ST. PAUL
80573        61702    Billing Dispute   BLOOMIGTON
13450        70711   Contract Dispute      CLIFTON
12102        77056      Debt Not Owed      HOUSTON
29136        77092  False Advertising      HOUSTON
44008        90010    Billing Dispute  LOS ANGELES
57636        92123         Harassment    SAN DIEGO
71001        92123    Billing Dispute    SAN DIEGO

Okay, there really are requests coming from LA and Houston! Good to know. Filtering by zip code is probably a bad way to handle this -- we should really be looking at the city instead.

In [16]:
requests['City'].str.upper().value_counts()

BROOKLYN               31662
NEW YORK               22664
BRONX                  18438
STATEN ISLAND           4766
JAMAICA                 2246
FLUSHING                1803
ASTORIA                 1568
RIDGEWOOD               1073
CORONA                   707
OZONE PARK               693
LONG ISLAND CITY         678
FAR ROCKAWAY             652
ELMHURST                 647
WOODSIDE                 609
EAST ELMHURST            562
QUEENS VILLAGE           549
JACKSON HEIGHTS          541
FOREST HILLS             541
SOUTH RICHMOND HILL      521
MASPETH                  473
WOODHAVEN                464
FRESH MEADOWS            435
SPRINGFIELD GARDENS      434
BAYSIDE                  411
SOUTH OZONE PARK         410
RICHMOND HILL            404
REGO PARK                402
MIDDLE VILLAGE           396
SAINT ALBANS             387
WHITESTONE               348
                       ...  
LYNBROOK                   1
ROSELYN                    1
JERSEY CITY                1
FREEPORT      

It looks like these are legitimate complaints, so we'll just leave them alone.

# 7.4 Putting it together

Here's what we ended up doing to clean up our zip codes, all together:

In [17]:
na_values = ['NO CLUE', 'N/A', '0']
requests = pd.read_csv('../data/311-service-requests.csv', 
                       na_values=na_values, 
                       dtype={'Incident Zip': str})

In [18]:
def fix_zip_codes(zips):
    # Truncate everything to length 5 
    zips = zips.str.slice(0, 5)
    
    # Set 00000 zip codes to nan
    zero_zips = zips == '00000'
    zips[zero_zips] = np.nan
    
    return zips

In [19]:
requests['Incident Zip'] = fix_zip_codes(requests['Incident Zip'])

In [20]:
requests['Incident Zip'].unique()

array(['11432', '11378', '10032', '10023', '10027', '11372', '11419',
       '11417', '10011', '11225', '11218', '10003', '10029', '10466',
       '11219', '10025', '10310', '11236', nan, '10033', '11216', '10016',
       '10305', '10312', '10026', '10309', '10036', '11433', '11235',
       '11213', '11379', '11101', '10014', '11231', '11234', '10457',
       '10459', '10465', '11207', '10002', '10034', '11233', '10453',
       '10456', '10469', '11374', '11221', '11421', '11215', '10007',
       '10019', '11205', '11418', '11369', '11249', '10005', '10009',
       '11211', '11412', '10458', '11229', '10065', '10030', '11222',
       '10024', '10013', '11420', '11365', '10012', '11214', '11212',
       '10022', '11232', '11040', '11226', '10281', '11102', '11208',
       '10001', '10472', '11414', '11223', '10040', '11220', '11373',
       '11203', '11691', '11356', '10017', '10452', '10280', '11217',
       '10031', '11201', '11358', '10128', '11423', '10039', '10010',
       '11209',

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  